edgedb-client/src/builder.rs

```diff
@@ -329,111 +350,96 @@ impl fmt::Display for DisplayAddr<'_> {
     }
 }
 
-impl DSN {
-    fn new(dsn: &str) -> Result<Self, Error> {
-        let admin = dsn.starts_with("edgedbadmin://");
-        if !dsn.starts_with("edgedb://") && !admin {
-            return Err(ClientError::with_message(format!(
-                "String {:?} is not a valid DSN",
-                dsn,
-            )));
-        };
-        let url = url::Url::parse(dsn).map_err(|e| {
-            ClientError::with_source(e).context(format!("cannot parse DSN {:?}", dsn))
-        })?;
-        let query = url
-            .query_pairs()
-            .into_owned()
-            .into_iter()
-            .collect::<HashMap<_, _>>();
+impl<'a> DsnHelper<'a> {
+    fn from_url(url: &'a url::Url) -> Result<Self, Error> {
+        let admin = url.scheme() == "edgedbadmin";
+        let mut query = HashMap::new();
+        for (k, v) in url.query_pairs() {
+            if query.contains_key(&k) {
+                return Err(ClientError::with_message(format!(
+                    "{k:?} is defined multiple times in the DSN query"
+                )));
+            } else {
+                query.insert(k, v);
+            }
+        }
         Ok(Self { url, admin, query })
     }
 
-    async fn get_value<T>(
+    async fn retrieve_value<T>(
         &mut self,
         key: &'static str,
-        mut v: Option<T>,
+        v: Option<T>,
```

Comment:
```suggestion
        v_from_url: Option<T>,
```
Since it's used twice there is no need to shorten it. And it's very unclear what is this value when reading function signature.